### DISCUSSION TITLE
Any tips for relocating SSE address offsets to VR edition?

### DISCUSSION THREAD

**Discussion by mysticfall** [alandtse/skyrim_vr_address_library#82] [Q&A] [answered]
Hi,

I've been trying to port [Skyrim Platform](https://www.nexusmods.com/skyrimspecialedition/mods/54909) to Skyrim VR using CommonLibSSE-NG.

I managed to make it compile with a VR configuration. However, it CTD because it fails to find an ID in the Address Library.

I believe this is because the mapping file shipped by VR Address Library for SKSEVR lacks entries for a handful of IDs defined in [this file](https://github.com/skyrim-multiplayer/skymp/blob/main/skyrim-platform/src/platform_se/skyrim_platform/game/Offsets.h).

If I understood it correctly, SSE and VR shares the ID so in theory, I can replace the "#ifdefs" from the source using CommonLibSSE-NG's relocation API, which I already did.

But the actual address offsets corresponding to those IDs must be present in the mappings database (e.g. `version-1-4-15-0.csv`) for it to work correctly at runtime.

However, I'm not a C++ programmer so reverse engineering binary files searching for matching addresses feels to be a daunting task.

Currently, offsets for these referenced IDs are missing:
```cpp
inline REL::Relocation<std::uintptr_t> SendEvent{ REL::ID(98077) };
inline REL::Relocation<std::uintptr_t> DrawSheatheWeaponPC{ REL::ID(40232) };
inline REL::Relocation<std::uintptr_t> DrawSheatheWeaponActor{ REL::ID(36289) };
inline REL::Relocation<std::uintptr_t> SendAnimation{ REL::ID(37020) };
inline REL::Relocation<std::uintptr_t> ApplyMasksToRenderTargets{ REL::ID(26454) };
inline REL::Relocation<std::uintptr_t> RenderCursorMenu{ REL::ID(32867) };

inline constexpr REL::ID WinMain(35545);

REL::Relocation<RE::ObjectRefHandle*> handle{ REL::ID(514164) };
REL::Relocation<func_t> func{ REL::ID(55622) };
REL::Relocation<func_t> func{ REL::ID(55682) }; // PushActorAway
```

I'm very willing to try, but I need some pointers about where should I start. Any kind of advice would be appreciated.

Thanks!

**Comment by alandtse** [comment, marked as answer]
Please take a look at the readme here. https://github.com/alandtse/vr_address_tools. It has some initial instructions. The issue is the IDs/addresses are the first step. You'll need to be able to review the assembly for both binaries to figure out if the code is the same and if not you may have to port/patch differently. I don't know what platform is doing, but if it's just hooking calls, that may be the easiest. If it's actually patching code, then you may have to learn assembly to perform the patch. 

Also, what version of Skyrim are they patching? The IDs changed with AE 1.6.x. VR matches SE 1.5.97 addresses.

**Reply by mysticfall** [reply]
> Please take a look at the readme here. https://github.com/alandtse/vr_address_tools. It has some initial instructions. The issue is the IDs/addresses are the first step. You'll need to be able to review the assembly for both binaries to figure out if the code is the same and if not you may have to port/patch differently. I don't know what platform is doing, but if it's just hooking calls, that may be the easiest. If it's actually patching code, then you may have to learn assembly to perform the patch.
> 
> Also, what version of Skyrim are they patching? The IDs changed with AE 1.6.x. VR matches SE 1.5.97 addresses.

Thanks for the answer.

Skyrim Platform is a TypeScript "binding" to SKSE, so to speak, so it usually just relays events and function invocations between the two sides. Also, because it's an SDK, we can just mark a few non-functioning ones as "not supported in VR yet" as long as critical APIs work ok.

As to VR Address Tools, I thought they were meant to analyse the source and identify ID references? I already made the necessary changes to the source, and all the missing address entries (i.e. missing matching entries in `database.csv`) are just those 10 lines that I listed in the above comment. I used CommonLibSSE-NG's relocation API to handle the differences in IDs between AE and SE/VR. (I'll use SE 1.5.97 since it's the edition I own. I can download this specific version using Steam console.)

The remaining task seems to be reviewing the assemblies to fill in those missing address entries. However, I have little clue about what tools or processes I should use to achieve the goal. Even general pointers would be of much help to me.

**Reply by alandtse** [reply]
You need to use ghidra or buy/pirate Ida. If you go ghidra, https://github.com/DaymareOn/SSE-Ghidra-Tutorial/wiki/SSE-Ghidra-Tutorial

**Reply by mysticfall** [reply]
> You need to use ghidra or buy/pirate Ida. If you go ghidra, https://github.com/DaymareOn/SSE-Ghidra-Tutorial/wiki/SSE-Ghidra-Tutorial

I'll definitely start by learning Ghidra, then. Thanks much!

**Reply by mysticfall** [reply]
> You need to use ghidra or buy/pirate Ida. If you go ghidra, https://github.com/DaymareOn/SSE-Ghidra-Tutorial/wiki/SSE-Ghidra-Tutorial

The link proved to be very helpful, and now I have managed to match most of the addresses I needed. I'll open a PR when I'm done and verify that they are working. Thanks again!
